File: frequencyTopic2020/linklist/addTwoNumbers.cc
#include <iostream>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(nullptr) {}
};

class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    if (l1 == nullptr) {
      return l2;
    }
    if (l2 == nullptr) {
      return l1;
    }
    int up = 0;
    ListNode *prev = new ListNode(0);
    ListNode *before = prev;
    while (l1 != nullptr || l2 != nullptr || up != 0) {
      int sum = up;
      if (l1 != nullptr) {
        sum += l1->val;
      }
      if (l2 != nullptr) {
        sum += l2->val;
      }
      if (sum >= 10) {
        up = sum / 10;
        sum = sum % 10;
      } else {
        up = 0;
      }
      ListNode *node = new ListNode(sum);
      prev->next = node;
      prev = node;
      if (l1 != nullptr){
        l1 = l1->next;
      }
      if (l2 != nullptr) {
        l2 = l2->next;
      }
    }
    return before->next;
  }
};

File: mock/ant/addTwoNumbers.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    prev := &ListNode{}
    dummy := prev
    var up int

    for l1 != nil || l2 != nil || up != 0 {
        sum := up
        if l1 != nil {
            sum += l1.Val
            l1 = l1.Next
        }
        if l2 != nil {
            sum += l2.Val
            l2 = l2.Next
        }
        if sum >= 10 {
            up = sum / 10
            sum %= 10
        } else {
            up = 0
        }
        node := &ListNode{Val: sum}
        prev.Next = node
        prev = prev.Next
    }

    return dummy.Next
}
